lib/refinement/analyzer.rb

```diff
@@ -0,0 +1,341 @@
+          end
+        end
+      end
+
+      scheme
+    end
+
+    # @return [String] a string suitable for user display that explains target changes
+    # @param include_unchanged_targets [Boolean] whether targets that have not changed should also be displayed
+    # @param change_level [Symbol] the change level used for computing whether a target has changed
+    def format_changes(include_unchanged_targets: false, change_level: :full_transitive)
+      annotate_targets!.group_by { |target| target.xcode_target.project.path.to_s }.sort_by(&:first)
+                       .map do |project, annotated_targets|
+        changes = annotated_targets.sort_by { |annotated_target| annotated_target.xcode_target.name }
+                                   .map do |annotated_target|
+          change_reason = annotated_target.changed?(level: change_level)
+          next if !include_unchanged_targets && !change_reason
+          change_reason ||= 'did not change'
+          "\t#{annotated_target.xcode_target}: #{change_reason}"
+        end.compact
+        "#{project}:\n#{changes.join("\n")}" unless changes.empty?
+      end.compact.join("\n")
+    end
+
+    private
+
+    def projects
+      @projects ||= find_projects(workspace_path)
+    end
+
+    def augmenting_paths_by_target
+      @augmenting_paths_by_target ||= begin
+        require 'yaml'
+        augmenting_paths_yaml_files.reduce({}) do |augmenting_paths_by_target, yaml_file|
+          yaml_file = Pathname(yaml_file).expand_path(changeset.repository)
+          yaml = YAML.safe_load(yaml_file.read)
+          augmenting_paths_by_target.merge(yaml) do |_target_name, prior_paths, new_paths|
+            prior_paths + new_paths
+          end
+        end
+      end
+    end
+
+    def map_find(enum)
+      enum.each do |elem|
+        transformed = yield elem
+        return transformed if transformed
+      end
+
+      nil
+    end
+
+    # @return [Array<AnnotatedTarget>] targets in the given list of Xcode projects,
+    #   annotated according to the given changeset
+    def annotated_targets
+      project_changes = Hash[projects.map do |project|
+        [project, project_referenced_in_changeset?(project: project)]
+      end]
+
+      require 'tsort'
+      targets = projects.flat_map(&:targets)
+      targets_by_uuid = Hash[targets.map { |t| [t.uuid, t] }]
+      targets_by_name = Hash[targets.map { |t| [t.name, t] }]
+      targets_by_product_name = Hash[targets.map do |t|
+        next unless t.respond_to?(:product_reference)
+        [File.basename(t.product_reference.path), t]
+      end.compact]
+
+      find_dep = ->(td) { targets_by_uuid[td.native_target_uuid] || targets_by_name[td.name] }
+      target_deps = lambda do |target|
+        a = []
+        target.dependencies.each do |td|
+          a << find_dep[td]
+        end
+
+        # TODO: also resolve OTHER_LDFLAGS?
+        # yay auto-linking
+        if (phase = target.frameworks_build_phases)
```

Comment:
does this even run? I thought it was `target.framework_build_phase` but I havent checked.

Reply:
It's definitely `frameworks_build_phases`, but you're right, I should add a test for resolving dependencies

Reply:
Interestingly enough the method is in plural but returns a single element.
```ruby
        def frameworks_build_phases
          build_phases.find { |bp| bp.class == PBXFrameworksBuildPhase }
        end
```

Oh well.